### DISCUSSION TITLE
Is there a way to stop actors system with discarding messages in the message boxes?

### DISCUSSION THREAD

**Discussion by svetlyak40wt** [mdbergmann/cl-gserver#88] [General]
I want to do a shutdown which will let actors finish their current tasks but will not let them to start new.

This is a small example of what I mean:

```lisp
(defun test-stopping (&key (wait t))
  (let* ((system (make-actor-system))
         (actor (flet ((func (msg)
                         (log:warn "Sleeping 10 seconds" msg)
                         (sleep 10)
                         (log:warn "Done")))
                  (actor-of system
                            :receive #'func
                            ;; :other-args '(:queue-size 1)
                            ))))
    (log:warn "Posting messages to the actor")
    (sento.actor:tell actor "first-time")
    (sento.actor:tell actor "second-time")
    (log:warn "Stopping the system")
    (sento.actor-context:shutdown system
                                  :wait wait)
    (log:warn "System was stopped")
    (values)))
```

Currently, this code will wait 20 seconds while actor will process both messages sent to it before the call to `shutdown` function. But I want to shutdown as soon as possible, discarding both or at least the last message, sent to the actor.

P.S. – I've tried to call `(sento.actor-cell:stop actor)` before shutdown, but it has no desired effect :(

**Comment by mdbergmann** [comment]
There are currently two ways how to stop an actor:

1. sending `:stop` to the actor. This will just enqueue this as a normal message and the actor is stopped once the message was processed.
2. executing `actor-cell:stop`. This disallows more message to be enqueued, but also 'just' normally enqueues a message that it will first need to process to be fully stopped. This is what `ac:shutdown` does.

Unfortunately your scenario is not implemented. I didn't need it yet, but it would be good to be there and it's not hard to do (I think).
I will provide a fix over the weekend (maybe today).

**Comment by svetlyak40wt** [comment]
Ok. Here is where I need such feature.

I'm building a cron-like addon for Sento. There is a macro defschedule, which can be used to define a schedule – right now it is a delay and a fbound symbol:

```lisp
(defun test-func ()
  (with-log-unhandled ()
    (let ((run-num (incf *run-num*))
          (num-seconds 120))
      (log:info "Sleeping" run-num num-seconds)
      (sleep num-seconds)
      (log:info "Sleeping Done" run-num))))

(defschedule *schedule*
  (30 test-func))
```

Also, there are two functions:

- `start-scheduler` - creates a new actor system, an actor for each scheduled function and uses `sento.wheel-timer:schedule-recurring` to schedule a helper which will `tell` an actor to do it's job.
- `stop-scheduler` - should stop everything and cleanup resources.

And here are two moments:

- an actor should have a limited to 1 queue size. Otherwise if job will require more time than a scheduling interval, the new messages will be stacked in the queue.
- shutdown does not work as I want and I have to wait while an actor will execute it's task a few times after shutdown was called.

Probably I'm overcomplicating and there is a simpler way to implement a cron-like system on actors?

**Reply by mdbergmann** [reply]
Well, usually the message processing of actors should not be blocked.
But it depends, in your case it is necessary.

Actor system comes with a scheduler, which can be used by actors. Here is an example:

```
ACTOR-CELL-TEST> (ac:actor-of *sys*
                              :init (lambda (self)
                                      (wt:schedule-recurring 
                                       (asys:scheduler (ac:system 
                                                        (act:context self)))
                                       1 1 (lambda () (act:! self "Foobar"))))
                              :receive (lambda (msg)
                                         (format t "msg: ~a~%" msg)))
msg: Foobar
msg: Foobar
msg: Foobar
msg: Foobar
```

This sets up a recurring timer in actors 'init' function. The timer function then just sends message to the actor which should do some action. This is to avoid race-condition on handling the timer function if actor state is involved.

The granularity of the scheduler is on a second basis.

Maybe this could be used instead of blocking message processing.

**Reply by svetlyak40wt** [reply]
Indeed! I'm using the similar code but call schedule-recurring not in the init.

What I'm talking about is that if you will add a `(sleep 5)` into the "receive" function, simulating some real job. Then queue of the actor will grow infinitely. And when you'll try to `shutdown` the system with `:wait t` argument, it will wait infinitely (until all items in the queue will be processed).

**Reply by mdbergmann** [reply]
Yeah, that's right.
I'm working on something to discard existing messages after 'stop'.

**Comment by svetlyak40wt** [comment]
Ok. I'll live without a stop function for now. Just wanted to let you know there is at least one more user of this feature :)

**Reply by mdbergmann** [reply]
Actually, if you try without `wait` it should already discard remaining messages, but doesn't wait until the current one is fully processed.

**Comment by mdbergmann** [comment]
So, please check the latest code. This should work now.

**Reply by svetlyak40wt** [reply]
No this works some-times, but somtimes doesn't. I've tested using the `test-stopping` function from my initial post.

Also, I've added some logging to the `stop` func (which you've modified in the latest commit https://github.com/mdbergmann/cl-gserver/commit/7445f1eaddcf56a0c3903618e3cfa3a06a66895d.

Here is a log from one of my runs. Pay attention to log records after the line `System was stopped`:

```
 <WARN> [23:56:10] sento-playground/playground slimeIZo4Ty (top level form test-stopping) -
  Posting messages to the actor
 <WARN> [23:56:10] sento-playground/playground slimeIZo4Ty (top level form test-stopping) -
  Stopping the system
 <WARN> [23:56:10] sento-playground/playground slimeIZo4Ty (top level form test-stopping func) - Sleeping 10 seconds SENTO-PLAYGROUND/PLAYGROUND::MSG: "first-time" 
<ERROR> [23:56:11] sento.messageb slimeaGRS6b (top level form stop message-box/bt) -
  SENDING tigger ending
<ERROR> [23:56:11] sento.messageb slimeaGRS6b (top level form stop message-box/bt) -
  SENDING tigger ending
<ERROR> [23:56:11] sento.messageb slimeaGRS6b (top level form stop message-box/bt) -
  SENDING tigger ending
<ERROR> [23:56:11] sento.messageb slimeaGRS6b (top level form stop message-box/bt) -
  SENDING tigger ending
 <WARN> [23:56:11] sento-playground/playground slimeIZo4Ty (top level form test-stopping) -
  System was stopped
; No values
 <WARN> [23:56:20] sento-playground/playground slimeIZo4Ty (top level form test-stopping func) - Done SENTO-PLAYGROUND/PLAYGROUND::MSG: "first-time" 
 <WARN> [23:56:20] sento-playground/playground slimeIZo4Ty (top level form test-stopping func) - Sleeping 10 seconds SENTO-PLAYGROUND/PLAYGROUND::MSG: "second-time" 
 <WARN> [23:56:30] sento-playground/playground slimeIZo4Ty (top level form test-stopping func) - Done SENTO-PLAYGROUND/PLAYGROUND::MSG: "second-time" 
```

The `second-time` messages is processed anyway even after the system stop.

However seldom during some `test-stopping` runs I see that `second-time` is not processed.

**Reply by svetlyak40wt** [reply]
Here is a trace of calls to actors and queues STOP functions.

```
   0 - sento.actor-cell:stop
     | > #<SENTO.ACTOR:ACTOR path: /user/test-stopping, cell: #<SENTO.ACTOR:ACTOR test-stopping, running: T, state: NIL, message-box: #<MESSAGE-BOX/DP mesgb-1463, processed messages: 2, max-queue-size: 0, queue: #<QUEUE-UNBOUNDED {700EAB53E3}>>>>
     | > NIL
     | < NIL
   1 `--- sento.messageb:stop
          > #<SENTO.MESSAGEB:MESSAGE-BOX/DP mesgb-1463, processed messages: 2, max-queue-size: 0, queue: #<QUEUE-UNBOUNDED {700EAB53E3}>>
          > NIL
   2 - sento.actor-cell:stop
     | > #<SENTO.ACTOR:ACTOR path: /internal/eventstream-actor-1461, cell: #<SENTO.ACTOR:ACTOR eventstream-actor-1461, running: T, state: NIL, message-box: #<MESSAGE-BOX/DP mesgb-1462, processed messages: 0, max-queue-size: 0, queue: #<QUEUE-UNBOUNDED {700EAB0333}>>>>
     | > NIL
     | < NIL
   3 `--- sento.messageb:stop
          > #<SENTO.MESSAGEB:MESSAGE-BOX/DP mesgb-1462, processed messages: 0, max-queue-size: 0, queue: #<QUEUE-UNBOUNDED {700EAB0333}>>
          > NIL
   4 - sento.actor-cell:stop
     | > #<SENTO.DISPATCHER:DISPATCH-WORKER path: /internal/dispatch(SHARED)-worker-4, cell: #<SENTO.DISPATCHER:DISPATCH-WORKER dispatch(SHARED)-worker-4, running: T, state: NIL, message-box: #<MESSAGE-BOX/BT mesgb-1460, processed messages: 0, max-queue-size: 0, queue: #<QUEUE-UNBOUNDED {700E9CD253}>>>>
     | > NIL
     | < NIL
   5 `--- sento.messageb:stop
          > #<SENTO.MESSAGEB:MESSAGE-BOX/BT mesgb-1460, processed messages: 0, max-queue-size: 0, queue: #<QUEUE-UNBOUNDED {700E9CD253}>>
          > NIL
          < NIL
   6 - sento.actor-cell:stop
     | > #<SENTO.DISPATCHER:DISPATCH-WORKER path: /internal/dispatch(SHARED)-worker-3, cell: #<SENTO.DISPATCHER:DISPATCH-WORKER dispatch(SHARED)-worker-3, running: T, state: NIL, message-box: #<MESSAGE-BOX/BT mesgb-1459, processed messages: 0, max-queue-size: 0, queue: #<QUEUE-UNBOUNDED {700E9CA2E3}>>>>
     | > NIL
     | < NIL
   7 `--- sento.messageb:stop
          > #<SENTO.MESSAGEB:MESSAGE-BOX/BT mesgb-1459, processed messages: 0, max-queue-size: 0, queue: #<QUEUE-UNBOUNDED {700E9CA2E3}>>
          > NIL
          < NIL
   8 - sento.actor-cell:stop
     | > #<SENTO.DISPATCHER:DISPATCH-WORKER path: /internal/dispatch(SHARED)-worker-2, cell: #<SENTO.DISPATCHER:DISPATCH-WORKER dispatch(SHARED)-worker-2, running: T, state: NIL, message-box: #<MESSAGE-BOX/BT mesgb-1458, processed messages: 0, max-queue-size: 0, queue: #<QUEUE-UNBOUNDED {700E9C72D3}>>>>
     | > NIL
     | < NIL
   9 `--- sento.messageb:stop
          > #<SENTO.MESSAGEB:MESSAGE-BOX/BT mesgb-1458, processed messages: 0, max-queue-size: 0, queue: #<QUEUE-UNBOUNDED {700E9C72D3}>>
          > NIL
          < NIL
  10 - sento.actor-cell:stop
     | > #<SENTO.DISPATCHER:DISPATCH-WORKER path: /internal/dispatch(SHARED)-worker-1, cell: #<SENTO.DISPATCHER:DISPATCH-WORKER dispatch(SHARED)-worker-1, running: T, state: NIL, message-box: #<MESSAGE-BOX/BT mesgb-1457, processed messages: 0, max-queue-size: 0, queue: #<QUEUE-UNBOUNDED {700CF4D823}>>>>
     | > NIL
     | < NIL
  11 `--- sento.messageb:stop
          > #<SENTO.MESSAGEB:MESSAGE-BOX/BT mesgb-1457, processed messages: 0, max-queue-size: 0, queue: #<QUEUE-UNBOUNDED {700CF4D823}>>
          > NIL
          < NIL
```

Here you can see my actor's name is `/user/test-stopping` and it's message box type is `MESSAGE-BOX/DP`, but in the latest commit you've modified  a `STOP` method for `MESSAGE-BOX/BT` class. Could this be a reason why stopping still does not work?

**Reply by mdbergmann** [reply]
I was thinking that it should be enough to only handle this in the /BT message-box because the /DP type of message box doesn't itself operate a thread, it just provides the function to execute on one of the dispatcher workers.
But maybe it's not enough and there also has to be some handling in the /DP message-box. I'll try to reproduce, but the automated tests were not able to catch this case.

**Reply by mdbergmann** [reply]
I've updated the code. Please give it a try.

**Reply by svetlyak40wt** [reply]
Cool! Now it works as expected.

By the way, did you make a force push? I see only one last commit with the fix.

**Reply by mdbergmann** [reply]
Awesome.
Yes, squashed and force pushed.

**Reply by svetlyak40wt** [reply]
It is not the best practice for something pushed to the `master` branch. Probably the next time we should work on some branch until the feature or fix will be done.

Anyway, thank you for a quick fix!

**Reply by mdbergmann** [reply]
Yes, that would be better. And a proper PR also and just squash in the end.

**Reply by svetlyak40wt** [reply]
Oh, you've reminded me that I've promised to create a PR with proposal, how to pass queue size to the actor's mailbox. I'll try to do this during the week.

**Comment by mdbergmann** [comment]
Message discarding implemented as part of 94a959185b4a1a9b7df907e8c6cf7b090d68dabc
